LVGL stage 2

clydebarrow opened the pull request:
# What does this implement/fix?

<!-- Quick description and explanation of changes -->

Adds touchscreens, and `btn` (allows testing touch input.)
Refactor widget typing.


DO NOT USE THIS PR as an external component - it is not a complete implementation. See #6363 for a working implementation of LVGL.



## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3678

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [x] RP2040
- [x] host

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml
lvgl:
  touchscreens: my_touchscreen
  log_level: TRACE
  bg_color: light_blue
  default_font: space16
  widgets:
    - btn:
        widgets:
          - label:
              text: Button
        align: top_mid
        pressed:
          bg_color: 0x00FFFF
    - label:
        text: Hello world
        text_color: 0xFF8000
        align: center
        text_font: montserrat_40
        border_post: true
        pressed:
          text_color: 0x00FFFF
    - label:
        text: "Hello shiny day"
        text_color: 0xFFFFFF
        align: bottom_mid

font:
  - file: "gfonts://Roboto"
    id: space16
    size: 40
    bpp: 4

color:
  - id: light_blue
    hex: "3340FF"
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

clydebarrow committed:
LVGL base implementation

clydebarrow committed:
CODEOWNERS

clydebarrow committed:
script/custom-ci

clydebarrow committed:
Remove esp8266 test

clydebarrow committed:
Newline

clydebarrow committed:
clang-tidy

clydebarrow committed:
Fix bug introduced by clang-tidy's suggestion.

clydebarrow committed:
Fix output of lv_scr_act()

clydebarrow committed:
clang-tidy fixes

clydebarrow committed:
clang-tidy

clydebarrow committed:
Fix fonts

clydebarrow committed:
Cleanup

clydebarrow committed:
Add test for font

clydebarrow committed:
CI fix

clydebarrow committed:
Add touchscreens

clydebarrow committed:
Add test images

clydebarrow committed:
Add tests for all obj styles

jesserockz committed:
`ruff check --fix`

jesserockz committed:
`ruff format`

clydebarrow committed:
ruff

clydebarrow committed:
ruff

clydebarrow committed:
Ruff

clydebarrow committed:
Merge branch 'dev' of https://github.com/esphome/esphome into lvgl-base

clydebarrow committed:
Review fixes

clydebarrow committed:
Fail hard when display mis-configured

clydebarrow committed:
Fix default_font for esphome font

clydebarrow committed:
Merge branch 'lvgl-base' into lvgl-input

clydebarrow committed:
Implement btn.

clydebarrow committed:
Add percent width

clydebarrow committed:
Review changes

clydebarrow committed:
Review changes

clydebarrow committed:
Properly handle display list

clydebarrow committed:
Properly handle display

clydebarrow committed:
Add btn.py

clydebarrow committed:
Merge branch 'dev' of https://github.com/esphome/esphome into lvgl-base

clydebarrow committed:
Improve display schema

clydebarrow committed:
Merge branch 'lvgl-base' into lvgl-input

clydebarrow committed:
Clean up touchscreen config

clydebarrow committed:
Fix import

clydebarrow committed:
Merge remote-tracking branch 'upstream/dev' into lvgl-input

clydebarrow committed:
Remove unused function

clydebarrow committed:
Use ESPHome image for test

brotherdust commented:
@clydebarrow , thank you for all your hard work on this! I'm trying this out on my LilyGo T4 S3 and am running into some compile errors with the config options. Specifically:
`
  [pages] is an invalid option for [lvgl]. Please check the indentation.
  pages: 
    - id: main_page
      widgets: 
        - label: 
            align: center
            text: Hello world!
`
If I remove pages, the widget by itself compiles and works fine on the device:
`
lvgl:
  displays:
    - disp_0
  touchscreens:
    - touch_0
  default_font: montserrat_24
  # pages:
  #   - id: main_page
  widgets:
  - label:
      align: center
      text: "Hello world!"
`

Okay I can live without pages. Let's try a slider?
`
  widgets: 
    - 
      [slider] is an invalid option for [widgets]. Please check the indentation.
      slider: 
        id: slider_test
        x: 20
        y: 50
        width: 30
        height: 220
`
I'm following the examples in the latest documentation. Maybe I'm doing it wrong? Either way, I understand this is not a complete implementation. Just wanted to bring this up in case anyone else runs into it.

Thanks!

nagyrobi commented:
@brotherdust please don't use this PR, as it only contains partial features. Use https://github.com/esphome/esphome/pull/6363 instead.

brotherdust commented:
> @brotherdust please don't use this PR, as it only contains partial features. Use #6363 instead.

Will do. Thanks!

clydebarrow committed:
Fix add_flag
Make single touchscreen added by default.

clydebarrow committed:
Ensure undefined fonts are found in id validation step

clydebarrow committed:
Issue warning if touchscreen configured but not used.
Fix font identifier code generation

clydebarrow committed:
Allow for case of ESPHOME_LOG_LEVEL_NONE

brotherdust commented:
@clydebarrow / @nagyrobi

I have a use-case for the color wheel  widget. If you don't already have a plan to implement it, I would like to contribute the implementation to this integration. Would you be open to that, or should I just wait?

Thanks!

nagyrobi commented:
It was dropped in lvgl9. Maybe it would be better to manually implement a color selector with sliders.

brotherdust commented:
> It was dropped in lvgl9. Maybe it would be better to manually implement a color selector with sliders.

I noted that as I was perusing the LVGL docs yesterday.

I take it, then, that moving to v9 is on the roadmap?

clydebarrow commented:
> I take it, then, that moving to v9 is on the roadmap?

Yes, on hold for now pending merge and release of current version and maturity of LVGL9. There are issues with functionality and performance ATM.

nagyrobi commented:
We'd like to avoid breaking changes on long term.
It's not yet decided what's the best approach:
a) implement lvgl8 now but just with the limited set of features wich allow a possibly clean (aka with no breaking changes) migration for users in ESPHome
b) implement lvgl8 now as complete as possible and later implement lvgl9 (and any future version) as a separate component, allowing the users to be prepsred for a manual migration with breaking changes at their will

Both have advantages and disadvantages. At the moment we're on a).

brotherdust commented:
Fair. Even on v8, my Lilygo T3 S4, which has an ESP32-S3, 8MB of SRAM and a quad-SPI display, seems to struggle a bit. With that in mind, the newer ESP32-P4 looks promising.

Thanks again for implementing this. The docs are high quality!

clydebarrow committed:
Rename btn to button, and img_ properties to image_

clydebarrow committed:
Rationalise conf_button

jesserockz commented on esphome/components/lvgl/lvgl_esphome.h:
```diff
@@ -18,12 +18,16 @@
 #ifdef USE_LVGL_FONT
 #include "esphome/components/font/font.h"
 #endif
+#ifdef LV_USE_TOUCHSCREEN
```
```suggestion
#ifdef USE_LVGL_TOUCHSCREEN
```

jesserockz commented on esphome/components/lvgl/lvgl_esphome.h:
```diff
@@ -107,12 +119,47 @@ class LvglComponent : public PollingComponent {
   bool full_refresh_{};
 };
 
+#ifdef LV_USE_TOUCHSCREEN
```
```suggestion
#ifdef USE_LVGL_TOUCHSCREEN
```

jesserockz commented on esphome/components/lvgl/lvgl_esphome.h:
```diff
@@ -107,12 +119,47 @@ class LvglComponent : public PollingComponent {
+  touchscreen::TouchPoint touch_point_{};
+  bool touch_pressed_{};
+};
+#endif
```
```suggestion
#endif  // USE_LVGL_TOUCHSCREEN
```

jesserockz commented on esphome/components/lvgl/touchscreens.py:
```diff
@@ -0,0 +1,38 @@
+        add_lv_use("TOUCHSCREEN")
+        touchscreen = await cg.get_variable(tconf[CONF_TOUCHSCREEN_ID])
+        lpt = tconf[CONF_LONG_PRESS_TIME].total_milliseconds & 0xFFFF
+        lprt = tconf[CONF_LONG_PRESS_REPEAT_TIME].total_milliseconds & 0xFFFF
```
If these have a max of uint16 max, then this should be in the validation.

jesserockz commented on esphome/components/lvgl/__init__.py:
```diff
@@ -93,8 +97,12 @@ def final_validation(config):
-    if not CORE.is_host and buffer_frac > 0.5 and "psram" not in global_config:
+    if CORE.is_esp32 and buffer_frac > 0.5 and "psram" not in global_config:
         LOGGER.warning("buffer_size: may need to be reduced without PSRAM")
+    if "touchscreen" in global_config and not config[df.CONF_TOUCHSCREENS]:
+        LOGGER.warning(
+            "Touchscreen not used by LVGL unless specified with lvgl `touchscreens:` key"
+        )
```
I feel like having multiple displays and multiple touchscreens is adding extra complexity to the code and extra configuration for a user.

If I was writing a YAML, and it automatically attaches to the display, then it should automatically attach to the touchscreen as well. Nobody is going to write a YAML file with a display, touchscreen and lvgl and not want to use the touchscreen.

Maybe we should remove multi-display/touch and can add it back later as an optional/advanced option. I don't see it being used in many real world situations anyway.

nagyrobi replied:
Physical display/touch + VNC display/touch for remote viewing/interact...

nagyrobi replied:
There can be setups without touch/rotary/keypad (just a status display).

clydebarrow replied:
Touchscreen now automatically used if a single one present, but not specifically configured in `lvgl:`. Multiples of both display and touchscreen still permitted.

clydebarrow committed:
Address review:
A single touchscreen now automatically picked up;
long press times validated to be <= 65535ms
ifdef/endifs corrected.

clydebarrow committed:
Merge branch 'dev' of https://github.com/esphome/esphome into lvgl-input

jesserockz commented on esphome/components/lvgl/touchscreens.py:
```diff
@@ -0,0 +1,46 @@
+        listener = cg.new_Pvariable(tconf[CONF_ID], lpt, lprt)
+        await cg.register_parented(listener, var)
+        lv.indev_drv_register(listener.get_drv())
+        lv_add(touchscreen.register_listener(listener))
```
Tested this locally and it is working with the normal cg instead of being placed inside the init lambda
```suggestion
        cg.add(touchscreen.register_listener(listener))
```

clydebarrow replied:
As you saw, I've done that, but it may have to be reverted in the future. There is a potential for touchscreen events to be sent to the listener before the the parent component has been completely setup. Right now I can't see that will cause any problems.

jesserockz reviewed (changes requested):
Looks good, just one minor thing that caught my eye =)

clydebarrow committed:
Move touchscreen listener add to main code
